### PR TITLE
feat: refresh token rotation configuration & github and facebook providers

### PR DESCRIPTION
**Pull request by davenewza**
### `facebook` and `gitlab` as built-in OpenID Connect providers

No need to provide issuer, token or auth URLs for Facebook and GitLab.  They only need to be specified like this in the config:

```yaml
  providers:
    - type: facebook
      name: fb_myapp
      clientId: 1234567
```

### Refresh Token Rotation configureable

For some clients it may be important to disable refresh token rotation.  When rotation is disabled, the original refresh token will be returned when minting a new access token.  This can be done in the config as follows:

```yaml
  tokens:
    accessTokenExpiry: 3600
    refreshTokenExpiry: 86400
    refreshTokenRotationEnabled: false. # <-- disable rotation
```

### Other changes

 - Improved reading from the config and dealing with default values
 - Cleaned up some existing tests which made no more sense

### CODE REVIEW

File: runtime/apis/authapi/token_endpoint.go [hunk 1]
```diff
@@ -66,6 +67,15 @@ func TokenEndpointHandler(schema *proto.Schema) common.HandlerFunc {
 		ctx, span := tracer.Start(r.Context(), "Token Endpoint")
 		defer span.End()
 
+		var identityId string
+		var refreshToken string
+
+		config, err := runtimectx.GetOAuthConfig(ctx)
+		if err != nil {
+			span.RecordError(err)
```
Comment (jonbretman) [inline runtime/apis/authapi/token_endpoint.go:75]: with stack trace?

File: runtime/apis/authapi/token_endpoint.go [hunk 2]
```diff
@@ -111,10 +121,24 @@ func TokenEndpointHandler(schema *proto.Schema) common.HandlerFunc {
...
-			isValid, newRefreshToken, identityId, err := oauth.RotateRefreshToken(ctx, refreshTokenRaw)
-			if err != nil {
-				span.RecordError(err)
-				return common.NewJsonResponse(http.StatusInternalServerError, nil, nil)
+			var isValid bool
+			if config.RefreshTokenRotationEnabled() {
+				// Rotate and revoke this refresh token, and mint a new one.
+				isValid, refreshToken, identityId, err = oauth.RotateRefreshToken(ctx, refreshTokenRaw)
+				if err != nil {
+					span.RecordError(err)
```
Comment (jonbretman) [inline runtime/apis/authapi/token_endpoint.go:129]: with stack trace?

File: runtime/apis/authapi/token_endpoint.go [hunk 3]
```diff
@@ -217,35 +226,37 @@ func TokenEndpointHandler(schema *proto.Schema) common.HandlerFunc {
...
 			return common.NewJsonResponse(http.StatusBadRequest, &TokenErrorResponse{
 				Error:            TokenEndpointUnsupportedGrantType,
 				ErrorDescription: "the only supported grants are 'refresh_token' and 'token_exchange'",
 			}, nil)
 		}
+
+		// Generate a new access token for this identity.
+		accessTokenRaw, expiresIn, err := oauth.GenerateAccessToken(ctx, identityId)
+		if err != nil {
+			span.RecordError(err)
```
Comment (jonbretman) [inline runtime/apis/authapi/token_endpoint.go:248]: why do you hate stack traces? 😄

File: config/auth.go [hunk 4]
```diff
@@ -40,6 +54,33 @@ type Provider struct {
 	AuthorizationUrl string `yaml:"authorizationUrl"`
 }
 
+// AccessTokenExpiryOrDefault retrieves the configured or default access token expiry
+func (c *AuthConfig) AccessTokenExpiryOrDefault() time.Duration {
```
Comment (jonbretman) [inline config/auth.go:58]: Kind of feel like the `OrDefault` suffix on these methods is not needed. It's explained in the doc comment, and typed to always return a `time.Duration`.

File: runtime/apis/authapi/token_endpoint.go (hunk 1 above)
Comment (davenewza) [reply runtime/apis/authapi/token_endpoint.go:78]: Oops yes.  I've also set `span.SetStatus(codes.Error, err.Error())` for these (only the "internal server" errors though)

File: runtime/apis/authapi/token_endpoint.go (hunk 3 above)
Comment (davenewza) [reply runtime/apis/authapi/token_endpoint.go:251]: I swear I really don't 😭

File: config/auth.go (hunk 4 above)
Comment (davenewza) [reply config/auth.go:58]: Sounds good - I've fixed them

**Pull request merged**